x-pack/elastic-agent/pkg/agent/configuration/settings.go

```diff
@@ -15,12 +15,13 @@ import (
 
 // SettingsConfig is an collection of agent settings configuration.
 type SettingsConfig struct {
-	DownloadConfig   *artifact.Config                `yaml:"download" config:"download" json:"download"`
-	ProcessConfig    *process.Config                 `yaml:"process" config:"process" json:"process"`
-	GRPC             *server.Config                  `yaml:"grpc" config:"grpc" json:"grpc"`
-	RetryConfig      *retry.Config                   `yaml:"retry" config:"retry" json:"retry"`
-	MonitoringConfig *monitoringCfg.MonitoringConfig `yaml:"monitoring" config:"monitoring" json:"monitoring"`
-	LoggingConfig    *logger.Config                  `yaml:"logging,omitempty" config:"logging,omitempty" json:"logging,omitempty"`
+	DownloadConfig        *artifact.Config                `yaml:"download" config:"download" json:"download"`
+	ProcessConfig         *process.Config                 `yaml:"process" config:"process" json:"process"`
+	GRPC                  *server.Config                  `yaml:"grpc" config:"grpc" json:"grpc"`
+	RetryConfig           *retry.Config                   `yaml:"retry" config:"retry" json:"retry"`
+	MonitoringConfig      *monitoringCfg.MonitoringConfig `yaml:"monitoring" config:"monitoring" json:"monitoring"`
+	LoggingConfig         *logger.Config                  `yaml:"logging,omitempty" config:"logging,omitempty" json:"logging,omitempty"`
+	InstrumentationConfig *InstrumentationConfig          `yaml:"instrumentation,omitempty" config:"instrumentation,omitempty" json:"instrumentation,omitempty"`
```

Comment:
I know that only a few settings that appear under the `agent.*` settings are written to `fleet.yml`, such as the `agent.monitoring` settings.
If we want to keep apm settings under `agent.instrumentation` we need to make sure that they can be read/written to `fleet.yml` as well.

Do we want to keep them under `agent.instrumentation` or move it under `agent.monitoring.`?
Another question is why `instrumentation` and not `apm` (the [docs](https://www.elastic.co/guide/en/apm/agent/go/current/configuration.html) seem to use `apm`)?

Reply:
I was trying to stay consistent with the fleet instrumentation PR, which has already [been merged](https://github.com/elastic/fleet-server/pull/898/files#diff-7a93ad3d35c4a6f46f0b58e67a5c63503a0a5eefcc7fe48d15147e7331a107f7). We didn't have a formal discussion about the naming, though. If nesting under `agent.instrumentation` presents problems (how difficult is it to make it read/writable to fleet.yml?), we could change both this and the other PR.

I personally don't want to nest it under `agent.monitoring` just because it seems like it's about configuring exposing internally collected metrics, rather than pushing traces somewhere. This might just be an arbitrary separation my brain is making and to others it makes perfect sense to be nested under `monitoring`.

@marclop @ruflin @simitt @axw do you have an opinion on the nesting and naming?

Reply:
Maybe a little bit of hysterical raisins. The term "instrumentation" was originally used because it originated in apm-server's config. i.e. we had

```yaml
apm-server:
  instrumentation:
    enabled: true
```

Meaning apm-server's own instrumentation is enabled. We thought it would be rather confusing to name it `apm-server.apm`. Later we moved the config to libbeat, so it would just work for other Beats too. At this point we didn't stop to rethink the config name. TBH I don't love "instrumentation" -- sounds almost as meaningful as "stuff", but on the other hand it _is_ a collection of different things: metrics, traces, errors...

I don't have a good idea of what it should be though. `agent.monitoring` as a heading sounds OK. @ruflin?

Reply:
I didn't stop to rethink the names, but it's something that crossed my mind, in the end I didn't pursue it further since it seems that `libbeat` uses the `instrumentation` term and that is the current working configuration setting, on the other I don't know what other name would be fitting, we're shipping more than traces, but `agent.tracing` seems to be a name that seems fitting, even if we ship more than traces?

Reply:
@michel-laterman The `fleet.yml` part you refer to is the one that allows a user to have a managed elastic-agent with some local configs? If yes, it is something I don't like too much. It is not that we should not allow a user for some configs to have a local setting too but the way we implemented it I think adds more complexity than needed. We likely should take a step back here in general and discuss our approach forward before we add more settings to it and put us in a corner.

For the naming I recommend to disonnect from what we have in apm-server and fleet-server. If we would start from scratch, what would we name it in Elastic-Agent? We can still do this even if apm-server and fleet-server have different namings as we have a translation layer in place and fleet-server could change configs at any time as long as Elastic-Agent keeps coping with it, as it is never run standalone.

Reply:
What about something like this?

 - introduce `agent.monitoring.traces: true|false` (defaults to false)
 - introduce `agent.monitoring.apm_server.{hosts,api_key,secret_token}`:

e.g.

```yaml
agent.monitoring:
    logs: true
    metrics: true
    traces: true # enables tracing with the APM Go agent, uses apm_server for tracer's transport
    apm_server:
        hosts:
          - https://apm-server.one.com:8200
          - https://apm-server.two.com:8200
        secret_token: a_secret_token
        api_key: or_an_api_key
```

That leaves room, in the future, for sending metrics via APM Server (which may be part of a future architecture).

Reply:
IMO it generally would perfectly fit under `agent.monitoring` as it is complementary to metrics and logs collection. _Maybe_ an even better name would be `agent.observability.apm`? Just throwing in ideas here. 

@ruflin if this PR should make it into 8.0, we need to know how to move forward with adding the config options for now. I suggest to document and label the addition of apm as `beta`, so that we do not lock us in to the config options if we are not certain they stick around like this in the future. I'd expect this to be mainly used internally for now for troubleshooting. 
Ideally we align the fleet-server config options after a decision has made on this PR.